src/ol/Feature.js

```diff
@@ -265,13 +266,24 @@ class Feature extends BaseObject {
    * single style object, an array of styles, or a function that takes a
    * resolution and returns an array of styles. To unset the feature style, call
    * `setStyle()` without arguments or a falsey value.
-   * @param {import("./style/Style.js").StyleLike} [style] Style for this feature.
+   * @param {import("./style/Style.js").StyleLike|import("./style/flat.js").FlatStyleLike} [style] Style for this feature.
    * @api
    * @fires module:ol/events/Event~BaseEvent#event:change
    */
   setStyle(style) {
-    this.style_ = style;
-    this.styleFunction_ = !style ? undefined : createStyleFunction(style);
+    let styleLike;
+    if (
+      !style ||
+      typeof style === 'function' ||
+      style instanceof Style ||
+      (Array.isArray(style) && style.every((s) => s instanceof Style))
```

Comment:
This prevents test breaking unnecessary cloning of style arrays which was introduced in layer styles in #13900.  The check could be moved to `toStyleLike` to fix the problem at source.